WordPress/Classes/Services/Store.swift

```diff
@@ -9,11 +9,75 @@ enum ProductRequestError: ErrorType {
     case InvalidProductPrice
 }
 
-protocol StoreFacade {
-    func getProductsWithIdentifiers(identifiers: Set<String>, success: [Product] -> Void, failure: ErrorType -> Void)
+class StoreKitTransactionObserver: NSObject, SKPaymentTransactionObserver {
+    static let instance = StoreKitTransactionObserver()
+    func paymentQueue(queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
+        for transaction in transactions {
+            StoreKitCoordinator.instance.processTransaction(transaction)
+        }
+    }
+}
+
+// This is a workaround for StoreCoordinator not being able to have a static
+// stored property since it's a generic class.
+struct StoreKitCoordinator {
+    static let instance = StoreCoordinator(store: StoreKitStore())
+}
+
+class StoreCoordinator<S: Store> {
+    let store: S
+    var pendingPayment: (Plan, Int)? = nil
+
+    init(store: S) {
+        self.store = store
+    }
+
+    func purchasePlan(plan: Plan, product: S.ProductType, forSite siteID: Int) {
+        precondition(plan.productIdentifier == product.productIdentifier)
+        // TODO: fail better if there's a pending payment
+        precondition(pendingPayment == nil)
+        pendingPayment = (plan, siteID)
+        store.requestPayment(product)
+    }
+
+    func processTransaction(transaction: SKPaymentTransaction) {
+        DDLogSwift.logInfo("[Store] Processing transaction \(transaction)")
+        guard transaction.transactionState == .Purchased else {
+            // Ignore other states for now
+            if transaction.transactionState == .Failed {
+                DDLogSwift.logInfo("[Store] Finishing failed transaction \(transaction)")
+                SKPaymentQueue.defaultQueue().finishTransaction(transaction)
+            } else {
+                DDLogSwift.logInfo("[Store] Ignoring transaction \(transaction)")
+            }
+            return
+        }
+        assert(transaction.payment.productIdentifier == pendingPayment?.0.productIdentifier)
+        guard let siteID = pendingPayment?.1,
+            let plan = pendingPayment?.0,
+            let service = PlanService(siteID: siteID, store: StoreKitStore()),
+            let receiptURL = NSBundle.mainBundle().appStoreReceiptURL,
+            let receipt = NSData(contentsOfURL: receiptURL)
+            else {
+                assertionFailure()
```

Comment:
It is, but I still haven't figured out in which cases would that happen. What I've seen in a few places is Apple [suggesting](https://forums.developer.apple.com/thread/7142) that we ask for user confirmation before calling `SKReceiptRefreshRequest`, which I find really confusing

> In the case that the item is not in the in_app array, the app can present an alert to the effect that there appears to be a problem with the receipt and ask the user whether to refresh the receipt. When the user agrees, then make the SKReceiptRefreshRequest